core/transaction_pool/impl/pool_moderator_impl.cpp

```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "transaction_pool/impl/pool_moderator_impl.hpp"
+
+using kagome::common::Buffer;
+
+namespace kagome::transaction_pool {
+
+  PoolModeratorImpl::PoolModeratorImpl(std::shared_ptr<SystemClock> clock,
+                                       SystemClock::Duration ban_for,
+                                       size_t expected_size)
+      : ban_for_{ban_for},
+        clock_{std::move(clock)},
+        expected_size_{expected_size} {}
+
+  void PoolModeratorImpl::ban(const common::Hash256 &tx_hash) {
+    banned_until_.insert({tx_hash, clock_->now() + ban_for_});
+    if (banned_until_.size() > expected_size_ * 2) {
+      while (banned_until_.size() > expected_size_) {
+        banned_until_.erase(banned_until_.begin());
+      }
+    }
+  }
+
+  bool PoolModeratorImpl::banIfStale(primitives::BlockNumber current_block,
+                                     const primitives::Transaction &tx) {
+    if (tx.valid_till > current_block) {
+      return false;
+    }
+    ban(tx.hash);
+    return true;
+  }
+
+  bool PoolModeratorImpl::isBanned(const common::Hash256 &tx_hash) const {
+    auto it = banned_until_.find(tx_hash);
+    if (it == banned_until_.end()) {
+      return false;
+    }
+    // if ban time is exceeded, the transaction will be removed from the list
+    // on next updateBan()
+    return it->second >= clock_->now();
+  }
+
+  void PoolModeratorImpl::updateBan() {
+    auto now = clock_->now();
+    std::list<Map::iterator> removed;
+    for (auto it = banned_until_.begin(); it != banned_until_.end(); it++) {
+      if (it->second < now) {
+        removed.push_back(it);
+      }
+    }
+    for (auto &it : removed) {
+      banned_until_.erase(it);
+    }
```

Comment:
All this algorithm can be:
```
auto it = banned_until_.begin();
while (it != banned_until_.end() {
  if (it->second < now) {
    it = banned_until_.erase(it);
  } else {
    ++it;
  }
}
```

Reply:
I think you're right. Seems like initially there was an intention to return removed, but it happened to be unnecessary, so now it may be rewritten.